### PR TITLE
[INLONG-10183][Sort] MongoDB connector support audit ID

### PR DESCRIPTION
**Pull request by XiaoYou201**
### Prepare a Pull Request
- [INLONG-10183][Sort] MongoDB connectors support audit ID

- Fixes https://github.com/apache/inlong/issues/10183

- parent issue: https://github.com/apache/inlong/issues/8803

### Motivation
MongoDB connector not support audit function. So this pr will make it support audit function.

### Modifications
Modify MongoDB connector source.when source receive data, it will send audit information at same time.Some implement are as follows:

Program will get the audit information from the incoming sql statement and pass it to the deserialization classes. During deserialization, the program will report the number of data received.

企业微信截图_e874eff6-387b-4dda-abd4-201fc3c5b3a2

### CODE REVIEW

File: inlong-sort/sort-flink/sort-flink-v1.15/sort-connectors/mongodb-cdc/src/main/java/org/apache/inlong/sort/mongodb/MongoDBTableSource.java [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+                        splitMetaGroupSize,
+                        splitSizeMB,
+                        metricOption);
+        source.metadataKeys = metadataKeys;
+        source.producedDataType = producedDataType;
+        return source;
+    }
+
+    @Override
+    public boolean equals(Object o) {
```
Comment (aloyszhang) [inline inlong-sort/sort-flink/sort-flink-v1.15/sort-connectors/mongodb-cdc/src/main/java/org/apache/inlong/sort/mongodb/MongoDBTableSource.java:274]: Should the `equals()` and `hashcode()` need to take `metricOption` into account?

Comment (XiaoYou201) [reply inlong-sort/sort-flink/sort-flink-v1.15/sort-connectors/mongodb-cdc/src/main/java/org/apache/inlong/sort/mongodb/MongoDBTableSource.java:280]: yeah ~， you are right， fixed it， thx~

**Review by aloyszhang — DISMISSED** [review]

File: inlong-sort/sort-flink/sort-flink-v1.15/sort-connectors/mongodb-cdc/src/main/java/org/apache/inlong/sort/mongodb/MongoDBTableSource.java [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+            Optional.ofNullable(pollAwaitTimeMillis).ifPresent(builder::pollAwaitTimeMillis);
+            Optional.ofNullable(heartbeatIntervalMillis)
+                    .ifPresent(builder::heartbeatIntervalMillis);
+            Optional.ofNullable(splitMetaGroupSize).ifPresent(builder::splitMetaGroupSize);
+            Optional.ofNullable(splitSizeMB).ifPresent(builder::splitSizeMB);
+
+            return SourceProvider.of(builder.build());
+        } else {
+            com.ververica.cdc.connectors.mongodb.MongoDBSource.Builder<RowData> builder =
+                    com.ververica.cdc.connectors.mongodb.MongoDBSource.<RowData>builder()
```
Comment (EMsnap) [inline inlong-sort/sort-flink/sort-flink-v1.15/sort-connectors/mongodb-cdc/src/main/java/org/apache/inlong/sort/mongodb/MongoDBTableSource.java:203]: MongoDBSource is imported already

Comment (XiaoYou201) [reply inlong-sort/sort-flink/sort-flink-v1.15/sort-connectors/mongodb-cdc/src/main/java/org/apache/inlong/sort/mongodb/MongoDBTableSource.java:203]: The import class is com.ververica.cdc.connectors.**mongodb.source**.MongoDBSource ,

and here is  com.ververica.cdc.connectors.**mongodb**.MongoDBSource  ~

File: inlong-sort/sort-flink/sort-flink-v1.15/sort-connectors/mongodb-cdc/src/main/java/org/apache/inlong/sort/mongodb/MongoDBConnectorDeserializationSchema.java [hunk 3]
```diff
@@ -0,0 +1,807 @@
...
+            case UPDATE:
+                // It’s null if another operation deletes the document
+                // before the lookup operation happens. Ignored it.
+                if (fullDocument == null) {
+                    break;
+                }
+                GenericRowData updateAfter = extractRowData(fullDocument);
+                updateAfter.setRowKind(RowKind.UPDATE_AFTER);
+                out = new MetricsCollector<>(out, sourceMetricData);
+                emit(record, updateAfter, out);
```
[the remark this answers is not in the document]
Comment (XiaoYou201) [reply inlong-sort/sort-flink/sort-flink-v1.15/sort-connectors/mongodb-cdc/src/main/java/org/apache/inlong/sort/mongodb/MongoDBConnectorDeserializationSchema.java]: fix it

File: inlong-sort/sort-flink/sort-flink-v1.15/sort-connectors/mongodb-cdc/src/main/java/org/apache/inlong/sort/mongodb/MongoDBTableSource.java [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+    // Mutable attributes
+    // --------------------------------------------------------------------------------------------
+
+    /** Data type that describes the final output of the source. */
+    protected DataType producedDataType;
+
+    /** Metadata that is appended at the end of a physical source row. */
+    protected List<String> metadataKeys;
+
+    private MetricOption metricOption;
```
Comment (EMsnap) [inline inlong-sort/sort-flink/sort-flink-v1.15/sort-connectors/mongodb-cdc/src/main/java/org/apache/inlong/sort/mongodb/MongoDBTableSource.java:97]: should be final

Comment (XiaoYou201) [reply inlong-sort/sort-flink/sort-flink-v1.15/sort-connectors/mongodb-cdc/src/main/java/org/apache/inlong/sort/mongodb/MongoDBTableSource.java]: fix it

File: inlong-sort/sort-flink/sort-flink-v1.15/sort-connectors/mongodb-cdc/src/main/java/org/apache/inlong/sort/mongodb/MongoDBTableSource.java [hunk 5]
```diff
@@ -0,0 +1,341 @@
...
+
+/**
+ * A {@link DynamicTableSource} that describes how to create a MongoDB change stream events source
+ * from a logical description.
+ *
+ * Copy from com.ververica:flink-connector-mongodb-cdc-2.3.0
+ */
+public class MongoDBTableSource implements ScanTableSource, SupportsReadingMetadata {
+
+    private static final Logger LOG = LoggerFactory.getLogger(MongoDBTableSource.class);
```
[the remark this answers is not in the document]
Comment (XiaoYou201) [reply inlong-sort/sort-flink/sort-flink-v1.15/sort-connectors/mongodb-cdc/src/main/java/org/apache/inlong/sort/mongodb/MongoDBTableSource.java]: delted it

File: inlong-sort/sort-flink/sort-flink-v1.15/sort-connectors/mongodb-cdc/src/main/java/org/apache/inlong/sort/mongodb/MongoDBConnectorDeserializationSchema.java [hunk 6]
```diff
@@ -0,0 +1,804 @@
...
+                checkNotNull(
+                        extractBsonDocument(
+                                value, valueSchema, MongoDBEnvelope.DOCUMENT_KEY_FIELD));
+        BsonDocument fullDocument =
+                extractBsonDocument(value, valueSchema, MongoDBEnvelope.FULL_DOCUMENT_FIELD);
+        switch (op) {
+            case INSERT:
+                GenericRowData insert = extractRowData(fullDocument);
+                insert.setRowKind(RowKind.INSERT);
+                emit(record, insert, new MetricsCollector<>(out, sourceMetricData));
```
Comment (aloyszhang) [inline inlong-sort/sort-flink/sort-flink-v1.15/sort-connectors/mongodb-cdc/src/main/java/org/apache/inlong/sort/mongodb/MongoDBConnectorDeserializationSchema.java:148]: check `sourceMetricData` nullable first

Comment (XiaoYou201) [reply inlong-sort/sort-flink/sort-flink-v1.15/sort-connectors/mongodb-cdc/src/main/java/org/apache/inlong/sort/mongodb/MongoDBConnectorDeserializationSchema.java]: fixed it, thx~

**Review by EMsnap — DISMISSED** [review]

**Review by dockerzhang — APPROVED** [review]

**Review by EMsnap — APPROVED** [review]

**Review by aloyszhang — APPROVED** [review]

**Pull request merged**